File: ios/Classes/LofeltImpl.swift
//
//  LofeltImpl.swift
//  lofelt_sdk
//
//  Created by VS on 8/8/23.
//

import Foundation
import LofeltHaptics

class LofeltImpl: Lofelt {
    private var haptics: LofeltHaptics?
    
    func initHaptics() throws {
        haptics = try? LofeltHaptics.init()
    }
    
    func deviceMeetsMinimumRequirements() throws -> Bool {
        return LofeltHaptics.deviceMeetsMinimumRequirement();
    }
    
    func load(data: String) throws {
        try haptics?.load(data);
    }
    
    func play() throws {
        try haptics?.play();
    }
    
    func stop() throws {
        try haptics?.stop();
    }
    
    func seek(time: Double) throws {
        try haptics?.seek(Float(time));
    }
    
    func loop(enabled: Bool) throws {
        try haptics?.loop(enabled)
    }
    
    func setAmplitudeMultiplication(factor: Double) throws {
        try haptics?.setAmplitudeMultiplication(Float(factor))
    }
    
    func setFrequencyShift(shift: Double) throws {
        try haptics?.setFrequencyShift(Float(shift))
    }
    
    func clipDuration() throws -> Double {
        let duration = haptics?.getClipDuration() ?? 0;
        return Double(duration)
    }
}

File: android/src/main/kotlin/com/example/lofelt_sdk/LofeltImpl.kt
package com.example.lofelt_sdk

import Lofelt
import android.content.Context
import com.lofelt.haptics.LofeltHaptics

class LofeltImpl(private val context: Context) : Lofelt {

    var haptics: LofeltHaptics? = null

    override fun initHaptics() {
        haptics = LofeltHaptics(context)
    }

    override fun deviceMeetsMinimumRequirements(): Boolean {
        return haptics?.deviceMeetsMinimumRequirements() ?: false
    }

    override fun load(data: String) {
        haptics?.load(data.toByteArray())
    }

    override fun play() {
        haptics?.play()
    }

    override fun stop() {
        haptics?.stop()
    }

    override fun seek(time: Double) {
        haptics?.seek(time.toFloat())
    }

    override fun loop(enabled: Boolean) {
        haptics?.loop(enabled)
    }

    override fun setAmplitudeMultiplication(factor: Double) {
        haptics?.setAmplitudeMultiplication(factor.toFloat())
    }

    override fun setFrequencyShift(shift: Double) {
    }

    override fun clipDuration(): Double {
        return haptics?.clipDuration?.toDouble() ?: 0.0
    }
}